### PR TITLE
Do not use rpm --specfile to query the subpackage names

### PR DESCRIPTION
**Pull request by marcus-h**
Using "rpm --specfile..." is "insecure", because during macro
expansion arbitrary code can be executed ("%(...)"). Instead,
use the helpers/spec_query script, which does not expand such
macros, to list the names of all subpackages.

Note: this might break some packages (e.g., if a %package section
is surrounded by an %ifarch statement, but our "uname -m" guess
was wrong)

Fixes: boo#938556

### CODE REVIEW

**Comment by scarabeusiv** [comment]
Actually, rpm prints all subpkgs even if they are arch depending from what I can see with the specfile option.

Maybe the best solution would be to simply ignore the conditions?

**Comment by marcus-h** [comment]
On 2017-08-16 01:10:58 -0700, Tomáš Chvátal wrote:
> Actually, rpm prints all subpkgs even if they are arch depending from what I can see with the specfile option.
> 
Yep. What I had in mind was something like

%ifnarch x86_64
%package foo
...
%endif

which would not be shown on a x86_64 machine.

> Maybe the best solution would be to simply ignore the conditions?
> 
I'm also fine with this (I just tried mimic the old rpm -q
--specfile...). I'll update the patch.

**Comment by marcus-h** [comment]
On 2017-08-16 14:09:14 +0200, Marcus Hüwe wrote:
> On 2017-08-16 01:10:58 -0700, Tomáš Chvátal wrote:
> > Maybe the best solution would be to simply ignore the conditions?
> > 
> I'm also fine with this (I just tried mimic the old rpm -q
> --specfile...). I'll update the patch.

Force pushed an updated version of patch: we now ignore all
%if* conditionals during parsing (advantage: we get the same
results on all arches).

File: helpers/spec_query [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        push @{$dtags->{$tag}}, \$spec->[-1];
+        next;
+      }
+      $seen{$tag} = 1;
+    }
+    push @$spec, $_;
+  }
+  close(F) || die("close: $!\n");
+  # make duplicate source/patch tags unique (such duplicates probably
+  # only occur in a "pathological" specfile...)
```
Comment (scarabeusiv) [inline helpers/spec_query]: Actually I think this should be fail reason, basically, if there are "multiple" same files/patches then we should not allow the submission.

**Comment by marcus-h** [comment]
On 2017-08-16 12:41:00 +0000, Tomáš Chvátal wrote:
> > +    next if /^\s*%(?:if|else|endif)/;
> +    chomp;
> +    if ($unique_sources && /^\s*((?:Source|Patch)\d*)\s*:(.*)$/i) {
> +      my $tag = lc($1);
> +      if ($seen{$tag}) {
> +        push @$spec, "$tag: $2";
> +        push @{$dtags->{$tag}}, \$spec->[-1];
> +        next;
> +      }
> +      $seen{$tag} = 1;
> +    }
> +    push @$spec, $_;
> +  }
> +  close(F) || die("close: $!\n");
> +  # make duplicate source/patch tags unique (such duplicates probably
> +  # only occur in a "pathological" specfile...)
> 
> Actually I think this should be fail reason, basically, if there are "multiple" same files/patches then we should not allow the submission.
> 
Should such a check be in the scope of the 20-files-present-and-referenced
script? From the "rpm perspective" having something like

%if %foo
Source1: foo
%else
Source1: bar
%endif

is perfectly ok... Maybe this is something for the spec cleaner/formatter?
What do you think?

**Comment by scarabeusiv** [comment]
It is okay from the perspective of the rpmbuild. But from our perspective where we ship the src.rpm and expect users to have it working it is confusing, as most of these are conditions for archs/etc. So it should always be unique i think. @DimStar77 what do you think?

**Comment by DimStar77** [comment]
> It is okay from the perspective of the rpmbuild. But from our perspective where we ship the src.rpm and expect users to have it working it is confusing, as most of these are conditions for archs/etc. So it should always be unique i think. @DimStar77 what do you think?

If I catch them, I decline such cases, as it makes, as you state, the src.rpm not usable beyond the original distro it was built on.

The basic rule: sources and patches are to be added to the .src.rpm unconditional (exception: NoSource:) - and only be handled conditionally in %prep (extract, patch)

**Comment by marcus-h** [comment]
On 2017-08-16 16:06:48 +0000, Tomáš Chvátal wrote:
> It is okay from the perspective of the rpmbuild. But from our perspective where we ship the src.rpm and expect users to have it working it is confusing, as most of these are conditions for archs/etc. So it should always be unique i think.
> 
Yes, I agree, for factory packages it should be always unique.
I just force pushed a new version of the patch. Changes:
- if --unique-sources is passed, fail if source/patch tags are not unique
- add --disambiguate-sources option (currently unused) to disambiguate
  source/patch tags (if you prefer, I can also drop this code)
- in 20-files-present-and-referenced, removed "|| cleanup_and_exit 1"
  from the spec_query call in order exit with a more useful error message

**Comment by scarabeusiv** [comment]
I would really just make unique-sources the only option, but it does not hurt to keep it there if you want to have it there.

**Comment by marcus-h** [comment]
On 2017-08-17 04:14:50 -0700, Tomáš Chvátal wrote:
> I would really just make unique-sources the only option, but it does not hurt to keep it there if you want to have it there.
> 
Hmm what do you exactly mean? Currently, --unique-sources is used
and --disambiguate-sources is unused.

**Comment by scarabeusiv** [comment]
Sorry, meant it in a way that disambiguate-sources are the only allowed one.

**Comment by marcus-h** [comment]
On 2017-08-17 04:37:20 -0700, Tomáš Chvátal wrote:
> Sorry, meant it in a way that disambiguate-sources are the only allowed one.
> 
Hmm I'm a bit confused now:) We have:

--unique-sources:       fail if sources are not unique (this option is used
                        by the 20-files-present-and-referenced script)
--disambiguate-sources: the former --unique-sources behavior (renumber
                        duplicate source/patch tags) (_not_ used atm)

**Comment by scarabeusiv** [comment]
So I've read it correctly on the first round! :D

Basically, your code is correct then I leave up to your decision if you keep the code for the "unused" part.

**Comment by marcus-h** [comment]
On 2017-08-17 12:22:35 +0000, Tomáš Chvátal wrote:
> So I've read it correctly on the first round! :D
> 
> Basically, your code is correct then I leave up to your decision if you keep the code for the "unused" part.
> 
Ok. Then let's keep it for now... the option is even documented:p
Maybe it will be useful in the future...

**Pull request merged**